doc/build-helpers/fetchers.chapter.md

```diff
@@ -23,66 +26,444 @@ The following table shows an overview of the differences:
+
+For example, consider the following recipe that only uses a fetcher:
 
 ```nix
+{ fetchurl }:
 fetchurl {
-  url = "http://www.example.org/hello-1.0.tar.gz";
-  hash = "sha256-lTeyxzJNQeMdu1IVdovNMtgn77jRIhSybLdMbTkf2Ww=";
-};
+  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.05/.version";
+  hash = "sha256-ZHl1emidXVojm83LCVrwULpwIzKE/mYwfztVkvpruOM=";
+}
 ```
 
 A common mistake is to update a fetcher’s URL, or a version parameter, without updating the hash.
 
 ```nix
+{ fetchurl }:
 fetchurl {
-  url = "http://www.example.org/hello-1.1.tar.gz";
-  hash = "sha256-lTeyxzJNQeMdu1IVdovNMtgn77jRIhSybLdMbTkf2Ww=";
-};
+  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.11/.version";
+  hash = "sha256-ZHl1emidXVojm83LCVrwULpwIzKE/mYwfztVkvpruOM=";
+}
 ```
 
 **This will reuse the old contents**.
-Remember to invalidate the hash argument, in this case by setting the `hash` attribute to an empty string.
+Remember to invalidate the hash argument.
+You can do this by setting the `hash` attribute to an empty string.
 
 ```nix
+{ fetchurl }:
 fetchurl {
-  url = "http://www.example.org/hello-1.1.tar.gz";
+  url = "https://raw.githubusercontent.com/NixOS/nixpkgs/23.11/.version";
   hash = "";
-};
+}
 ```
 
-Use the resulting error message to determine the correct hash.
+When building the recipe, you'll get an error message that you can use to determine the correct hash:
 
-```
-error: hash mismatch in fixed-output derivation '/path/to/my.drv':
+```shell
+$ nix-build
+(some output removed for clarity)
+error: hash mismatch in fixed-output derivation '/nix/store/7yynn53jpc93l76z9zdjj4xdxgynawcw-version.drv':
          specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
-            got:    sha256-lTeyxzJNQeMdu1IVdovNMtgn77jRIhSybLdMbTkf2Ww=
+            got:    sha256-BZqI7r0MNP29yGH5+yW2tjU9OOpOCEvwWKrWCv5CQ0I=
+error: build of '/nix/store/bqdjcw5ij5ymfbm41dq230chk9hdhqff-version.drv' failed
+```
+
+A similar problem arises while testing changes to a fetcher's implementation.
+If the output of the derivation already exists in the Nix store, test failures can go undetected.
+The [`invalidateFetcherByDrvHash`](#tester-invalidateFetcherByDrvHash) function helps prevent reusing cached derivations.
+
+[]{#fetchurl}
+## `fetchurl` {#sec-pkgs-fetchers-fetchurl}
+
+Downloads content from a given URL and stores the unaltered contents within the Nix store.
+
+`fetchurl` uses {manpage}`curl(1)` internally, and allows its behaviour to be modified by specifying a few attributes in the argument to `fetchurl` (see the documentation for attributes `curlOpts`, `curlOptsList`, and `netrcPhase`).
+
+### File naming {#sec-pkgs-fetchers-fetchurl-file-naming}
+
+The name of the file within the Nix store is controlled by the hash given to `fetchurl`, and also the `name` (or `pname` and `version`) values.
```

Comment:
> The name of the file within the Nix store

I believe the official term is "derivation name"

Reply:
The Nix manual uses "derivation name" for the symbolic part of the name (which in the case of `fetchurl` would be determined by the `name` or `pname` + `version` attributes). I wanted to refer to the file name in this case, but the naming gets very confusing very fast (see below), so I prefer to stick to "name of the file within the Nix store" for now. At some point I want to attach a glossary to the Nixpkgs manual to make it easier to refer to names and their definitions, and at that point I hope I can start clarifying definitions in other manuals as well.

Why it gets very confusing very fast:
- In theory what I want to refer should be called the store path, but the [Nix manual glossary](https://nixos.org/manual/nix/stable/glossary#gloss-store-path) has a definition for store path which is not coherent with the [store path section](https://nixos.org/manual/nix/stable/store/store-path) in the same manual.
- Calling it a "file within the Nix store" is also not entirely true because there can be Nix stores that aren't part of the filesystem, but I couldn't find any better term that is easily understandable for the main audience of the manual.

Given that most people reading the manual will think of the Nix store as the one in their local filesystem, I chose to use that. The people who are aware of other store types and its quirks should be able to understand that the "file" part doesn't matter and this section is talking about the name of the object in the store more generally.

Reply:
ah, here you are referring to hash+name, but i believe earlier you used the same wording to describe just the derivation name?

Reply:
It was later when describing the `name`, `pname`, and `version` attributes. I just pushed a change to be more specific on what kind of "name" I'm talking about, while trying as much as possible to avoid more confusion.

Reply:
```suggestion
The resulting store path is determined by the hash given to `fetchurl`, and also the `name` (or `pname` and `version`) values.
```

And maybe use the Nix manual link from the line below, then we don't need an entire extra sentence.

Reply:
I split this PR into two. The `fetchurl` changes are in #300429 where I applied this suggestion.